draft-ietf-quic-tls.md

```diff
@@ -808,26 +807,110 @@ packet number limit.  An endpoint MUST initiate a key update ({{key-update}})
 prior to exceeding any limit set for the AEAD that is in use.
 
 
-## Packet Number Protection {#pn-encrypt}
+## Header Protection {#header-protect}
 
-QUIC packet numbers are protected using a key that is derived from the current
-set of secrets.  The key derived using the "pn" label is used to protect the
-packet number from casual observation.  The packet number protection algorithm
-depends on the negotiated AEAD.
+Parts of QUIC packet headers, in particular the Packet Number field, are
+protected using a key that is derived separate to the packet protection key and
+IV.  The key derived using the "quic hp" label is used to provide
+confidentiality protection for those fields that are not exposed to on-path
+elements.
 
-Packet number protection is applied after packet protection is applied (see
-{{aead}}).  The ciphertext of the packet is sampled and used as input to an
-encryption algorithm.
+This protection applies to the least-significant bits of the first byte, plus
+the Packet Number field.  The four least-significant bits of the first byte are
+protected for packets with long headers; the five least significant bits of the
+first byte are protected for packets with short headers.  For both header forms,
+this covers the reserved bits and the Packet Number Length field; the Key Phase
+bit is also protected for packets with a short header.
 
-In sampling the packet ciphertext, the packet number length is assumed to be 4
-bytes (its maximum possible encoded length), unless there is insufficient space
-in the packet for sampling.  The sampled ciphertext starts after allowing for a
-4 byte packet number unless this would cause the sample to extend past the end
-of the packet.  If the sample would extend past the end of the packet, the end
-of the packet is sampled.
+This process does not apply to Retry or Version Negotiation packets, which do
+not contain a protected payload or any of the fields that are protected by this
+process.
 
-For example, the sampled ciphertext for a packet with a short header can be
-determined by:
+
+### Header Protection Application
+
+Header protection is applied after packet protection is applied (see {{aead}}).
+The ciphertext of the packet is sampled and used as input to an encryption
+algorithm.  The algorithm used depends on the negotiated AEAD.
+
+The output of this algorithm is 5 bytes of mask which are applied to the
+protected using exclusive OR.  The least significant bits of the first byte of
+the packet are masked by the first mask byte, and the packet number is masked
+with the remaining bytes.
+
+{{pseudo-hp}} shows a sample algorithm for applying header protection. Removing
+protection only differs in the order in which the packet number length
+(pn_length) is determined.
+
+~~~
+mask = header_protection(hp_key, sample)
```

Comment:
Shouldn't this pseudo-code define how `sample` is computed first? The spec only defines it further down (in `#hp-sample`) so we may want to reorder the pseudo-code to somewhere after that